### PR TITLE
WebGLRenderer: Integrate RoughnessMipmapper.

### PR DESCRIPTION
**Pull request by Mugen87**
Related issue: see https://github.com/mrdoob/three.js/pull/22178#issuecomment-885781849

**Description**

This PR integrates `RoughnessMipmapper` into the core. All PBR materials with a roughness and normal map are now processed by this class in order to improve the rendering quality. One can see the subtle differences by comparing the `dev` and `prod` version of `webgl_tonemapping`. You can see on the left part of the helmet less aliasing/noise.

| dev | prod |
|---|---|
| image | image |

Since `RoughnessMipmapper`  was only used in one example, I did not add it to the public namespace. It should be a pure internal thing now.

@elalish I've moved a few checks from `RoughnessMipmapper.generateMipmaps()` to the new method `WebGLTextures.updateRoughnessMap()` so it's easier to manage the conversion process.

### CODE REVIEW

**Comment by WestLangley** [comment]
>You can see on the left part of the helmet less aliasing/noise.

Hmm. I see different noise. It would be great if @elalish can provide a use case where the benefit is more clear.

@elalish [said](https://github.com/mrdoob/three.js/pull/18110#issuecomment-564236033)

>The main difference is seen as reduced aliasing when zooming out.

**Comment by elalish** [comment]
Yeah, not very clear in that example. In general the benefits are more noticeable on materials that are shiny, have a fairly high-frequency normal map and have strong lighting (and of course angled to get the glint) and are not accessing the top level mip. Spruit sunrise tends to show it better (stronger HDR than this Venice shot). Zippers were a big driver for the feature when I first developed it. Also, the aliasing this removes tends to be strongly temporal, so it's not as noticeable in a still image as the interactive scene. Sorry that I don't have a good example handy; I'll look around.

File: src/renderers/webgl/WebGLTextures.js [hunk 1]
```diff
@@ -1372,9 +1375,42 @@ function WebGLTextures( _gl, extensions, state, properties, capabilities, utils,
 
 	}
 
+	function updateRoughnessMap( material ) {
+
+		const { roughnessMap, normalMap } = material;
+
+		// ignore PBR materials without roughness and normals maps
+
+		if ( roughnessMap === null || normalMap === null || roughnessMap.generateMipmaps === false ) return;
```
Comment (elalish) [inline src/renderers/webgl/WebGLTextures.js:1384]: Is it possible for the normal and roughness maps to use different UV coordinates? If so, we should disable this then too.

Comment (Mugen87) [reply src/renderers/webgl/WebGLTextures.js:1384]: They both use the first set of texture coordinates and the same texture transform.

**Comment by ManishJu** [comment]
I do not see much difference until you heighten in 10 times
this is the difference between the 2 files:
diff

this is the difference heightened 10 times:
diff2


difference heigtened 100 times:
diff3

**Comment by Mugen87** [comment]
Um, what do you exactly mean with "heighten"?

**Comment by ManishJu** [comment]
> 
> 
> Um, what do you exactly mean with "heighten"?

Multiplying the pixel intensity values by the number specified...

**Comment by mrdoob** [comment]
Sees like model-viewer is having a few reports of black renders (ie. https://github.com/google/model-viewer/issues/1955#issuecomment-927986115) and I'm wondering if the RoughnessMipmapper code is at fault...

Probably better to wait a bit before merging this.

**Pull request closed without merging**